[ADD] payment_nuvei: Add Nuvei payment provider integration

andg-odoo opened the pull request:
Nuvei is an online payment platform for the LATAM area. This integration
currently supports Argentina, Brazil, Canada, Chile, Colombia, Ecuador,
Mexico, Peru, United State, and Uruguay.

Nuvei supports card transactions as well as other local payment methods
unique per country.

task-4172642

See also:
- https://github.com/odoo/documentation/pull/11157

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/181459.png)](https://mergebot.odoo.com/odoo/odoo/pull/181459)

Demesmaeker commented on addons/payment_nuvei/const.py:
```diff
@@ -0,0 +1,74 @@
+}
+
+# The keys of the values to use in the calculation of the signature.
+INCOMING_KEYS = [
```
```suggestion
SIGNATURE_KEYS = [
```

Demesmaeker commented on addons/payment_nuvei/models/payment_transaction.py:
```diff
@@ -0,0 +1,205 @@
+        # Nuvei expects time sent in UTC/GMT
+        timestamp = self.create_date.strftime('%Y-%m-%d.%H:%M:%S')
+
+        # If the customer is buying something with Webpay in Chile, they expect USD transactions
+        # to be in whole integer values. Therefore we must round them.
+        if self.payment_method_code in const.INTEGER_METHODS:
+            rounding = 0
+        else:
+            rounding = self.currency_id.decimal_places
+
+
+        if self.payment_method_code in const.FULL_NAME_METHODS and not (first_name and last_name):
+            raise UserError(
+                "Nuvei: " + _(
+                    "%(payment_method)s requires both a first and last name.",
+                    payment_method=self.payment_method_code.capitalize()
+                )
+            )
+
+        rounded_amount = float_round(self.amount, rounding, rounding_method='DOWN')
```
So... If the buyer has to pay 1,5 USD... he will pay 1 USD instead? '^^ _(If this work for CLP, which decimal place is 0, Webpay also works with USD)_
Shouldn't we use `to_minor_currency_units` somewhere instead?
Also, why just this payment method? I didn't find where in the doc it was specified, and I find it strange that the same API would expect a different way to send the amount depending on the payment method code :thinking:

andg-odoo replied:
So, this one is fun. Talking with Nuvei, this is an issue with Webpay and their integration with it. For all of the non-card payment methods, their api is sending the data to the 3d-party providers in the correct format and it is reproducable in their system that the Webpay integration will error if you give it a USD amount with decimal amount. 

Talking with them, they could implement a rounding system on their end to handle it but the sprint planning system they have will take it a while (multi months) so ANV recommended I follow the flow that Mercado Pago and Xendit does in these two PRs:
- Xendit: https://github.com/odoo/odoo/pull/187525
- Mercado Pago: https://github.com/odoo/odoo/pull/185664

It does mean that the buyer will pay less than necessary, but I guess the hope is that the for Chilean customers they will have their products in whole numbers?

Demesmaeker commented on addons/payment_nuvei/__init__.py:
```diff
@@ -0,0 +1,14 @@
+from . import controllers
+from . import models
+
+from odoo.addons.payment import setup_provider, reset_payment_provider
```
Nitpicking:
```suggestion
from odoo.addons.payment import reset_payment_provider, setup_provider
```
Because: `ord('r') < ord('s')`

Demesmaeker commented on addons/payment_nuvei/controllers/main.py:
```diff
@@ -0,0 +1,121 @@
+    @http.route(
+        _cancel_url, type='http', auth='public', methods=['GET'], csrf=False, save_session=False
+    )
+    def nuvei_return_from_canceled_checkout(self, tx_ref, return_access_tkn, **data):
```
You can place this method before the webhook one, as it'll happen before if it happens

Demesmaeker replied:
And can't we, in fact, use the same base return url for both the cancel and the error, and if we find the tx_ref and return_access_tkn are also there, then we use these instead to search for the tx? It'd avoid some duplicate code and make it easier to understand.

Demesmaeker commented on addons/payment_nuvei/controllers/main.py:
```diff
@@ -0,0 +1,121 @@
+        if not payment_utils.check_access_token(return_access_tkn, tx_ref):
+            raise Forbidden()
+
+        if data:
+            tx_sudo._handle_notification_data('nuvei', data)
+        else:
+            # The customer canceled the transaction
+            tx_sudo._handle_notification_data('nuvei', {})
```
```suggestion
        tx_sudo._handle_notification_data('nuvei', data or {})
```

Demesmaeker reviewed (changes requested):
Hello @andg-odoo 

First review pass, not finished yet, but I'll already post it. Well done in general, there aren't any big changes required.

Could you please rewrite the commit and PR message, as this is the first thing people will read when they'll want explanation on some changes.Redirecting them to the readme isn't really practical, as it'll requires reading multiple files in the commit/PR. So reading the diff isn't often a good explanation on why a change was necessary, it only shows how. _(The readme doesn't explain why we decided to add this provider, and that's normal)_

Have a very nice day!

andg-odoo commented on addons/payment/data/payment_method_data.xml:
```diff
@@ -3731,6 +3820,15 @@
         <field name="image" type="base64" file="payment/static/img/bank.png"/>
     </record>
 
+    <record id="payment_method_banco_guayaquil" model="payment.method">
+        <field name="name">Banco Guayaquil</field>
+        <field name="code">banco_guayaquil</field>
+        <field name="primary_payment_method_id" eval="ref('payment.payment_method_astropay')"/>
+        <field name="sequence">1000</field>
+        <field name="active">False</field>
+        <field name="image" type="base64" file="payment/static/img/banco_guayaquil.png"/>
+    </record>
```
This is something I need to talk with the Ecuador PO here in the US because I'm not sure AstroPay is a familiar enough name in the country for online bank transactions or just the generic 3rd-party integration Nuvei picked. 

If it's the latter, would it be better to rename AstroPay to something like LATEM Bank Transactions since it's theoretically usable in more than Ecuador?

masi-odoo replied:
Hello @Demesmaeker @andg-odoo I was able to review this with the Ecuador PO and the "AstroPay" term is not that used or common in the country, so it would also be confusing for a final customer of the eCommerce.

We agree the best solution is to show the icons as they appear now, since the final client will be able to see terms that will be common like Banco Guayaquil, Banco Pichincha, and Facilito.

![image](https://github.com/user-attachments/assets/19ba58ae-9633-436b-849a-8a2622f4f4ad)

Demesmaeker replied:
Fine by me then, thanks for the explanation!

andg-odoo commented:
@Demesmaeker Thanks for the review, I have adjusted everything to your recommendations besides the one comment on AstroPay TEF.

You were right that combining the controllers worked and I was able to put everything within the return controller and simplify the module.

The commit message has also been updated as has the PR description, let me know if there is anything else necessary :)

Demesmaeker reviewed (changes requested):
Hello @andg-odoo 
Thanks for your changes ^^
While testing, I had a question. I could see that setting at least an item in the order seems to be mandatory, but it leaves us with this:
![image](https://github.com/user-attachments/assets/2905656f-5195-47d0-9329-644e96676086)
I think it's quite strange to have a quantity for an order. Isn't it possible to hide this somehow?
Otherwise, everything seems to be working well. And aside the check for the access token, these aren't critical changes.
Have a very nice day,

andg-odoo commented on addons/payment_nuvei/README.md:
```diff
@@ -0,0 +1,34 @@
+(You must match the card number and cardholder name to what is listed in frictionless/etc depending
+on what you are testing and then follow the expiration date/security code information from below)
+
+### VISA (Up to $99)
```
For certain countries that number works great for values over 100. But I've found randomly mexico and other countries fail silently with no reason using this card, so to be safe it's easier to have them split into two different cards.

andg-odoo commented on addons/payment_nuvei/controllers/main.py:
```diff
@@ -0,0 +1,102 @@
+        if ref and token:
+            # We only enter this flow on errors or a canceled order because Nuvei doesn't send
+            # the checksum on these cases.
+            if not payment_utils.check_access_token(token, ref):
```
That makes sense, I think using the secret key is something that makes sense since we use that to hash and verify the other checksum so it's good to match.

andg-odoo commented on addons/payment_nuvei/models/payment_transaction.py:
```diff
@@ -0,0 +1,200 @@
+        if self.partner_zip:
+            url_params['zip'] = self.partner_zip
+        if self.partner_state_id.code:
+            url_params['state'] = self.partner_state_id.code
```
I double checked with Nuvei and it should be supported in both test and production. I originally had them separate due to an issue with the hashing of the checksum but since then that problem has been solved.

andg-odoo commented on addons/payment/static/img/abitab.png:
These two were unused but they just enabled the final payment provider last thursday so I pushed it today. (It's server is down currently and they say they are fixing it)

Demesmaeker reviewed (changes requested):
Hello @andg-odoo 

Thanks for your changes, here are some lasts small nitpickings changes for the business code.
Meanwhile, for your tests, can I ask you to already follow the pattern of this task: https://www.odoo.com/odoo/project/4106/tasks/3054481
Most of your tests will stay the same, and some are obviously specific to nuvei, but a lot of these tests are missing and could be added.

Have a very nice week,

andg-odoo commented:
@Demesmaeker thank you for the review, I've adjusted the code to your suggestions and I added the tests that logically make sense to me but I might be missing some still.

I've skipped all the `test_reference_` cases because the Nuvei provider doesn't modify the reference at all unlike APS or AsiaPay.

I've also skipped all of the tests related to refunds, tokenization, or capture as it isn't implemented with the module at this stage.

Let me know if I missed one but it should be better now.

Demesmaeker commented on addons/payment_nuvei/tests/common.py:
```diff
@@ -0,0 +1,50 @@
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.maxDiff = None
```
Why is this necessary?

andg-odoo replied:
I had it when doing local debugging on the tests to see what was different. It is removed now.

Demesmaeker commented on addons/payment_nuvei/tests/common.py:
```diff
@@ -0,0 +1,50 @@
+
+        cls.notification_data = {
+            "ppp_status": "OK",
+            "LifeCycleId": "7110000000004858226",
```
You can only set the values that are necessary. We never check this `LifeCycleId` for instance, so no need to add it in the test setup

Demesmaeker commented on addons/payment_nuvei/tests/test_payment_provider.py:
```diff
@@ -0,0 +1,47 @@
+    def test_incompatible_with_unsupported_currencies(self):
+        """ Test that Nuvei providers are filtered out from compatible providers when the currency
+        is not supported. """
+        compatible_providers = self.env['payment.provider']._get_compatible_providers(
+            self.env.company.id, self.partner.id, self.amount, currency_id=self.env.ref('base.AFN').id
+        )
```
```suggestion
        currency_id = self.env.ref('base.AFN').id
        compatible_providers = self.env['payment.provider']._get_compatible_providers(
            self.env.company.id, self.partner.id, self.amount, currency_id=currency_id
        )
```
Nit to stay < 100 chars

Demesmaeker commented on addons/payment_nuvei/tests/test_payment_transaction.py:
```diff
@@ -0,0 +1,146 @@
+@tagged('post_install', '-at_install')
+class TestPaymentTransaction(NuveiCommon):
+
+    def _get_expected_values(self, tx):
```
As this is used only once, to test the rendering values, might be better to not extract it and instead let it inside the `test_no_item_missing_from_rendering_values` for clarity

Demesmaeker commented on addons/payment_nuvei/tests/test_payment_transaction.py:
```diff
@@ -0,0 +1,146 @@
+class TestPaymentTransaction(NuveiCommon):
+
+    def _get_expected_values(self, tx):
+        return_url = self._build_url(NuveiController._return_url)
+        first_name, last_name = payment_utils.split_partner_name(tx.partner_name)
+
+        webhook_url = self._build_url(NuveiController._webhook_url)
+        cancel_url_params = {
+            'tx_ref': tx.reference,
+            'return_access_tkn': self._generate_test_access_token(
+                self.provider.nuvei_secret_key,
+                tx.reference,
+            ),
+        }
+
+        phone_number = tx.partner_phone and tx._phone_format(
+            number=tx.partner_phone, country=tx.partner_country_id, raise_exception=False
+        )
```
Don't use other methods to get the expected values in the check. If one of these methods is changed and sent something wrong, the test would still be green as both the test values and the check values would have the same problem. Instead, hardcode the expected values whenever possible.

andg-odoo replied:
I had to unhardcode `self._generate_test_access_token` because since it uses `database.secret` in the hmac function, it kept changing everytime a new build was pushed. For other things, they are staying hardcoded.

Demesmaeker replied:
FYI: another solution would be to patch that method, as it isn't the one that's tested here =]
I'm fine with both anyway, so you don't have to change it again

Demesmaeker commented on addons/payment_nuvei/tests/test_payment_transaction.py:
```diff
@@ -0,0 +1,146 @@
+        ):
+            processing_values = tx._get_specific_rendering_values(None)
+        expected_values = self._get_expected_values(tx)
+        self.assertDictEqual(processing_values['url_params'], expected_values)
```
What about the `api_url` and `checksum`? You can compare all the rendering values, not just the `url_params` ^^

Demesmaeker commented on addons/payment_nuvei/tests/test_payment_transaction.py:
```diff
@@ -0,0 +1,146 @@
+
+    @mute_logger('odoo.addons.payment.models.payment_transaction')
+    def test_no_input_missing_from_redirect_form(self):
+        """ Test that the no key is not omitted from the rendering values. """
```
```suggestion
        """ Test that the no key is omitted from the rendering values. """
```

Demesmaeker commented on addons/payment_nuvei/tests/test_payment_transaction.py:
```diff
@@ -0,0 +1,146 @@
+        tx._process_notification_data(self.notification_data)
+        self.assertEqual(tx.state, 'done')
+
+
```
```suggestion
```

Demesmaeker commented on addons/payment_nuvei/tests/test_payment_transaction.py:
```diff
@@ -0,0 +1,146 @@
+        input_keys.sort()
+        self.assertListEqual(input_keys, expected_input_keys)
+
+    def test_processing_notification_data_confirms_transaction(self):
```
You can also test with the other states that are implemented in your `_process_notification_data`, for instance:
- `test_processing_notification_data_sets_transaction_in_error`

Demesmaeker commented on addons/payment_nuvei/tests/test_processing_flows.py:
```diff
@@ -0,0 +1,94 @@
+        self._create_transaction(flow='redirect')
+        url = self._build_url(NuveiController._return_url)
+        with patch(
+            'odoo.addons.payment_nuvei.controllers.main.NuveiController._verify_notification_signature'
```
```suggestion
            'odoo.addons.payment_nuvei.controllers.main.NuveiController'
            '._verify_notification_signature'
```

Demesmaeker commented on addons/payment_nuvei/tests/test_processing_flows.py:
```diff
@@ -0,0 +1,94 @@
+        self._create_transaction('redirect')
+        url = self._build_url(NuveiController._webhook_url)
+        with patch(
+            'odoo.addons.payment_nuvei.controllers.main.NuveiController._verify_notification_signature'
```
same

Demesmaeker reviewed (approved):
Hello @andg-odoo 
Thanks again for your changes, last nitpickings and good for me. I'll already green it as none of these changes would impact the incomming testing.
Have a very nice day!

andg-odoo commented on addons/payment_nuvei/__manifest__.py:
```diff
@@ -0,0 +1,19 @@
+
+        'data/payment_provider_data.xml',
+    ],
+    'post_init_hook': 'post_init_hook',
+    'uninstall_hook': 'uninstall_hook',
+    'license': 'LGPL-3',
```
I removed it because the style error showed up on runbot saying it was duplicating the default value and complained. If that can just be overriden on merge I'll add it back or add it in the FW-port to master.

andg-odoo commented:
@Demesmaeker I really appreciate you taking the time to look so in depth on the PR! I have made the changes requested, added a few more tests, and added back the author key which will trigger the style error but I assume we can override it later?

ticodoo commented on addons/payment_nuvei/i18n/payment_nvuei.pot:
So close, but looks like your pot file name has a typo. Needs to be fixed or the sync will break

Demesmaeker replied:
Nice catch!!

andg-odoo replied:
I don't know how I missed that, thanks @ticodoo! The new name has been pushed.

AntoineVDV commented on addons/payment_nuvei/views/payment_nuvei_templates.xml:
```diff
@@ -0,0 +1,13 @@
+    <template id="redirect_form">
+        <form t-att-action="api_url" method="post">
+            <input type="hidden" name="checksum" t-att-value="checksum" />
+            <t t-foreach="url_params" t-as="param">
```
Good initiative! It's true that we usually stick to listing every param but, unless their value should be computed, there is no reason to do that.

andg-odoo replied:
I'm a little confused on what you are recommending. Are you saying rather than having the view contain all of the inputs we just have one input containing the dictionary of everything?

All the data returned from the `_get_specific_rendering_values` call needs to be sent as post data so I'm not sure another way to have a form and post it all. 

I'm probably missing something obvious though.

AntoineVDV replied:
I'm not recommending anything, just highlighting that it was a good initiative :grinning:

AntoineVDV commented on addons/payment_nuvei/tests/test_payment_transaction.py:
```diff
@@ -0,0 +1,176 @@
+        tx._process_notification_data(payload)
+        self.assertEqual(
+            tx.state_message,
+            'An error occurred during the processing of your payment (Invalid Card).'
```
This is a bit overkill in a test. You want to test for logic that might break inadvertently, not for changes that would always be conscious and require to also adapt the test for no added value.

AntoineVDV reviewed (changes requested):
Hello, thank you both for your great work!

My review comments below are mostly nitpicking so I already pushed a fixup commit to resolve most of them.

Please fix the remaining issues and go through that commit, then squash it if it's all good :slightly_smiling_face: (don't rebase unless needed)

andg-odoo commented on addons/payment_nuvei/models/payment_transaction.py:
```diff
@@ -0,0 +1,194 @@
+
+        first_name, last_name = payment_utils.split_partner_name(self.partner_name)
+
+        # Nuvei expects time sent in UTC/GMT
```
In their docs they state: 
> The GMT time (in 24-hour format) that the transaction took place in the following format

But I thought they were interchangable terms. Realize now one is a timezone and one more of a concept.

andg-odoo commented:
@AntoineVDV The only changes in the fixup are:
- Removed Uruguay from AstroPay since Local Payments covers more + everything covered in AstroPay.
- Changed `payment_method_code.capitalize()` -> `payment_method_id.name`
- Added back the `user_token_id` parameter and a test case with it but generates it completely randomly since we don't want tokenization.

The `user_token_id` is not an obvious change, the reason for it is that Boleto, Pix, and WebPay (All Brazil and Chile providers) require it even if tokenization is disabled. Since we don't want to store the payment methods, the solution I had was just to pass a UUID4 string every time to make it unique. 

I don't like the solution but Nuvei doesn't have a way around this for now, hopefully it isn't for long if we decide to add tokenization as a second PR in the future.

AntoineVDV commented on addons/payment_nuvei/models/payment_transaction.py:
```diff
@@ -0,0 +1,188 @@
+            ),
+            'phone1': phone_number or '',
+            'state': self.partner_state_id.code or '',
+            'user_token_id': uuid4(),
```
```suggestion
            'user_token_id': uuid4(),  # Random string because required by some PMs but not used.
```

AntoineVDV commented on addons/payment_nuvei/tests/test_payment_transaction.py:
```diff
@@ -0,0 +1,168 @@
+
+    def test_no_item_missing_from_rendering_values(self):
+        """ Test that the rendering values match what we expect. """
+        tx = self._create_transaction(flow='redirect')
+        def make_uuid():
+            return "0000-0000-0000-0000"
```
```suggestion
        def make_uuid():
            return "0000-0000-0000-0000"
            
        tx = self._create_transaction(flow='redirect')
```

AntoineVDV reviewed (approved):
All good! You can squash your commits and we'll merge on Monday after we disable the false positive ci/template check due to merging a new module in stable :)

AntoineVDV commented:
@robodoo r+ override=ci/security
Reason:
- `csrf=False`: HTTP route opting out CSRF protection because called from external service and authenticated with signature computed based on shared secret.
- `hashlib`: Used to compute requests signature based on shared secret, and not on the DB secret.

Link to ci/security build (18.0): https://runbot.odoo.com/runbot/build/76612427

C3POdoo commented:
Upgrade exception [#529](https://runbot.odoo.com/odoo/runbot.upgrade.exception/529?menu_id=405) added.

Waiting the forward-port of this pr, the exception will be applied on all builds, and create an inconsistent state for migrations.
**Please forward port this asap up to master without change.** Exception should be forward-ported before the end of the week.

PRs adding fields/models should be merged at the beginning of the week and all the forward ports should be merged in the same week.

If you need to apply any change before it reaches master, please notify runbot team.

Details:
```
module:payment_nuvei
```
cc @KangOl @nseinlet @aj-fuentes

masi-odoo commented:
Thanks @andg-odoo @AntoineVDV !
